### PR TITLE
[26428] Custom fields in projects table

### PR DESCRIPTION
**Pull request by wielinde**
https://community.openproject.com/wp/26428

- [x] Migrate existing filter to generic filter mechanism
- [x] Show filter for project custom fields
- [x] Show selected custom fields in table
- [x] Show user custom values in table
- [x] Rename "Status" to "Active or archived"
- [x] Sort filter in form alphabetically
- [x] CF filter only for EE
- [x] Preserve sort and filter criteria (sort, pagination, filter)
- [x] Respect visibility settings of custom field
- [x] ~~Move enterprise token check into the custom field filte~~ decided against it as I don't see a reason to limit the API that way 
- [x] Fix tests
- [x] Extract custom field filter into mixin
- [x] Remove copied code from filters
- [x] Fix selected filter not disabled in add filter select until value added
- [x] Be able to remove the status filter
- [x] Accessibility
  - [x] Cannot reach the context menu
  - [x] Cannot reach the expand icon
- [x] Implement custom field sorting
- [x] fix problems with subproject indicator
- [x] ~~Have context menu trigger have stable position regardless of whether there is a description to expand~~ not important enough
- [x] Fix code climate issues
- [x] specify and implement default sorting
- [x] communicate EE features in CE

**Out of scope - to be done later**
- Rename attribute from "visibble" to "private" and create an additional permission to view "private" project custom values
- X-Icon top right to close the filter section in addition to pressing the filter button again (needs to be changed also in other filter)
- Show the pagination buttons only if needed. In case all projects fit on one screen these buttons are not needed.
- Modal to select columns
- URL targets for custom list values

### CODE REVIEW

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/607b17b75f2d641267e5b1bd2809632674474270) (deployed commit: 08a3665f34c02058dbafc20a5f2837c69608dc6e).

**Comment by lindenthal** [comment]
Hi @wielinde,

thanks for this great piece of software. 

I just added some remarks to the description. Please feel free to extract these task to later PRs.

Best
Niels

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/4804215d47ffed7a52efcdf39c88a89d672e1460) (deployed commit: 08a3665f34c02058dbafc20a5f2837c69608dc6e).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e1030d548a7a4ce817c0c5a3e98755aee029fcb2) (deployed commit: 08a3665f34c02058dbafc20a5f2837c69608dc6e).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/966d214ba5070dcdec9b2d63157031d8041709df) (deployed commit: 5639fcb4c3135ffcd1fabea6b258395e466e6e5e).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9604f8ca729013577c180260203348e7bc6640e0) (deployed commit: e0409b0c636fb87be52abfe8f8bd8c3478011f04).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/05ac5dcd23b28ae9be3ec84638a6ec25500da560) (deployed commit: eed0b5abb76d77967e2111ce2771a5996f9b28b6).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/95f55521413c19e75a0351173ca19a3801dfc9e7) (deployed commit: 49d61e120ebc875d2709d24722932e080871a3d3).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/21c7fb68d95ae1406d5c65626772b7e99e2ec1bf) (deployed commit: 49d61e120ebc875d2709d24722932e080871a3d3).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e46565d9407d8277c9701eea5462a7837ee7688e) (deployed commit: f2b689ab8caa090277f0750837901c406b54679a).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/92e9377601f03711f27ff014b29a7f9948b117e2) (deployed commit: f2b689ab8caa090277f0750837901c406b54679a).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/9101d88baf30f43419ece15daeed675f641b5682) (deployed commit: f2b689ab8caa090277f0750837901c406b54679a).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/8dfc607cf1cd88aff2634d4e4b408362320c33d4) (deployed commit: f2b689ab8caa090277f0750837901c406b54679a).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/2520ad90db57d0325b546075cd97a2afd28591c4) (deployed commit: e17ddb7a35a5bfef8b3660d9af633bed92e7558f).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ae148c73bd0b249d491af1a5f1dc9e00ff01dbb8) (deployed commit: e17ddb7a35a5bfef8b3660d9af633bed92e7558f).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/6a0c2e95460bf998045f546c959e59332882f954) (deployed commit: ae46ed554b8821aa70486d1064fe775ecca45eb7).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/2cbb5f59ccbe8743f6ef94c78ef68795b67a5cd2) (deployed commit: 7f8fe5ff2269c4397ddd90a341d805ae2747fed2).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/40428acceee719cb1367098d6e22cf76cae15949) (deployed commit: 7f8fe5ff2269c4397ddd90a341d805ae2747fed2).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/70a5b63ea341b8d9720eec470272afd5fb45cb1e) (deployed commit: 04433f5aa2a203462a0b26726563d72978b81a69).

File: app/models/queries/base_query.rb [hunk 1]
```diff
@@ -87,11 +89,17 @@ def default_scope
     self.class.model.all
   end
 
+  def find_active_filter(name)
+    filters.index_by(&:name)[name]
+  end
+
+  def find_available_filter(name)
+    available_filters.index_by(&:name)[name]
```
Comment (ulferts) [inline app/models/queries/base_query.rb]: Cool, didn't know the method `index_by`

As cool as it is, it is a bit excessive to create a new Hash to then fetch but one entry from it. consider using `detect`

```
available_filters.detect { |f| f.name == name }
```

File: app/models/queries/filters/strategies/base_strategy.rb [hunk 2]
```diff
@@ -62,6 +66,14 @@ def supported_operator_classes
         .sort_by { |o| self.class.supported_operators.index o.symbol.to_s }
     end
 
+    def default_operator_class
+      if self.class.default_operator
+        operator_map[self.class.default_operator]
+      else
+        operator_map[self.class.available_operators.first]
+      end
```
Comment (ulferts) [inline app/models/queries/filters/strategies/base_strategy.rb]:
```
operator = self.class.default_operator || self.class.available_operators.first
operator_map[operator]
```
is equivalent and shorter

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/d478e90074ba9da2189a89636e535671ea0e0965) (deployed commit: 259dec6c9acee79536d94c951be185757d31335d).

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -298,21 +297,30 @@ def add_current_user_to_project_if_not_admin(project)
...
+        .with_required_storage
+        .with_latest_activity
+        .order(sort_clause)
+        .page(page_param)
+        .per_page(per_page_param)
+
+      filter_projects_by_permission projects
+    else
+      flash[:error] = @query.errors.full_messages
+      return false
```
Comment (ulferts) [inline app/controllers/projects_controller.rb]: the `return` is unnecessary.

File: app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -325,6 +333,15 @@ def filter_projects_by_permission(projects)
     end
   end
 
+  def get_custom_fields
+    custom_fields = CustomField.where(type: 'ProjectCustomField')
```
Comment (ulferts) [inline app/controllers/projects_controller.rb]:
```
custom_fields = ProjectCustomField.all
```
is equivalent and less prone to break.

And as there is basically only one if statement in the method, the whole of the method could be written as:

```
if User.current.admin?
  ProjectCustomField.all
else
  ProjectCustomField.where(visible: true)
end
```

Seeing this, it would make sense to move the method to project_custom_field.rb:
```
def self.visible(user = User.current)
  if user.admin?
    all
  else
    where(visible: true)
  end
```
And then always call
```
ProjectCustomField.visible(User.current)
```
in the controller.

File: app/helpers/projects_helper.rb [hunk 5]
```diff
@@ -125,10 +125,50 @@ def format_version_sharing(sharing)
   end
 
   def filter_set?
-    if params[:status] || params[:name]
+    if params[:filters].present?
       true
     else
       false
     end
```
Comment (ulferts) [inline app/helpers/projects_helper.rb]:
```
def filter_set?
  if params[:filters].present?
    true
  else
    false
  end
```

= 

```
def filter_set?
  params[:filters].present?
end
```

File: app/helpers/projects_helper.rb [hunk 6]
```diff
@@ -125,10 +125,50 @@ def format_version_sharing(sharing)
...
   end
+
+  def allowed_filters(query)
+    filters_static = %i(status name_and_identifier created_on latest_activity_at)
+    filters_dynamic = []
+
+    if EnterpriseToken.allows_to?(:custom_fields_in_projects_list)
+      filters_dynamic_query = ProjectCustomField.where("field_format <> 'version'")
```
Comment (ulferts) [inline app/helpers/projects_helper.rb]: ProjectCustomField.where.not(field_format: 'version')

File: app/helpers/projects_helper.rb [hunk 7]
```diff
@@ -125,10 +125,50 @@ def format_version_sharing(sharing)
...
+
+      unless User.current.admin?
+        filters_dynamic_query = filters_dynamic_query.where(visible: true)
```
Comment (ulferts) [inline app/helpers/projects_helper.rb]: reuse `visible` scope from above after the code has been moved to the model

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/93bbb73d2e8dc20e130d82c80fcab62eff9fe014) (deployed commit: 4d6b282d8512fe8a019a2fbfdbca66dbd62a9b2d).

File: app/helpers/projects_helper.rb [hunk 8]
```diff
@@ -125,10 +125,50 @@ def format_version_sharing(sharing)
...
+    unless User.current.admin?
+      filters = filters - admin_only_filters
+    end
+
+    filter_instances = filters.map do |name|
+      query.find_available_filter(name)
+    end
+
+    filter_instances.sort_by(&:human_name)
+  end
```
Comment (ulferts) [inline app/helpers/projects_helper.rb]: I've been thinking about where to put this code. And it probably boils down to whether one wants to have the restrictions on the filters that is implemented here for the UI to also apply to the API. Given the code current implementation, the custom field filters and the filters for latest_activity_at and created_on would be accessible via the API which might or might not be desirable. 

If the limitations should also apply to the API, parts of the code would have to be placed in the filter (e.g. the limitation on the `ProjectCustomField.visible`)

On top of that, one could probably make use of the registration to get the filters programatically. Depending in parts on the question raised above, one would then have to narrow down the set of all possible filters here or on a more general level:

```
def allowed_filters(query)
  all_possible_filter_classes = query.class.registered_filters
  all_possible_filter_classes -=  blacklisted_project_filters
  all_possible_filter_classes.map(&:all_for).sort_by(&:human_name)
end

def blacklisted_project_filters
   blacklist = [Queries::Projects::Filters::AncestorFilter]
   blacklist += [Queries::Projects::Filters::CreatedOnFilter, Queries::Projects::Filters::LatestActivityAtFilter] unless User.current.admin?
   blacklist << Queries::Projects::Filters::CustomFieldFilter unless EnterpriseToken.allows_to?(:custom_fields_in_projects_list)
   blacklist
end

File: app/helpers/sort_helper.rb [hunk 9]
```diff
@@ -240,8 +240,8 @@ def sort_link(column, caption, default_order, html_options = {})
 
     sort_options = { sort: @sort_criteria.add(column.to_s, order).to_param }
 
-    # relying on url_for to take the rest of the current params from the request
-    link_to_content_update(h(caption), sort_options, html_options)
+    # Don't lose other params.
+    link_to_content_update(h(caption), request.query_parameters.merge(sort_options), html_options)
```
Comment (ulferts) [inline app/helpers/sort_helper.rb]: Careful: This way, the user can provide information which will then be embedded into the page. E.g. open this url   http://localhost:3000/projects?host=blubs and take a look at the url that will be used for sorting the project table headers. In short, this adds a vulnerability to XSS and DoS attacks to the application. You'd need to whitelist the accepted parameters to circumvent the problem

File: app/models/queries/filters/strategies/date.rb [hunk 10]
```diff
@@ -28,8 +28,10 @@
 #++
 
 module Queries::Filters::Strategies
-  class Date < Queries::Filters::Strategies::Integer
+  class Date < BaseStrategy
```
Comment (ulferts) [inline app/models/queries/filters/strategies/date.rb]: doing this removes the validation(integer) from the filter. Generally speaking, the calls to super will not do what was expected.

File: app/models/queries/filters/strategies/date_time_past.rb [hunk 11]
```diff
@@ -28,15 +28,19 @@
...
     def validate
       if operator == Queries::Operators::OnDateTime ||
          operator == Queries::Operators::BetweenDateTime
         validate_values_all_datetime
-      else
-        super
+      elsif operator == Queries::Operators::Ago ||
+            operator == Queries::Operators::LessThanAgo ||
+            operator == Queries::Operators::MoreThanAgo
+        validate_values_all_integer
```
Comment (ulferts) [inline app/models/queries/filters/strategies/date_time_past.rb]: There are two operators now, that are not validated any more. What is the reason for this change. Why not stick with inheriting from Queries::Filters::Strategies::Integer?

File: app/models/queries/projects/filters/custom_field_filter.rb [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+
+  def where_joins(db_table, project_db_table)
+    joins = "LEFT OUTER JOIN #{db_table}
+               ON #{db_table}.customized_type='Project'
+               AND #{db_table}.customized_id=#{project_db_table}.id
+               AND #{db_table}.custom_field_id=#{custom_field.id}"
+
+    joins
+  end
+end
```
Comment (ulferts) [inline app/models/queries/projects/filters/custom_field_filter.rb]: This is copied from the work package`s custom field filter. Please extract both into a module then included into both classes.

File: app/models/queries/projects/filters/name_and_identifier_filter.rb [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+  def where_not_contains(values)
+    conditions = []
+    assignments = []
+    values.each do |value|
+      conditions << "LOWER(projects.identifier) LIKE ? OR LOWER(projects.name) LIKE ?"
+      assignments = assignments + ["%#{value.downcase}%", "%#{value.downcase}"]
+    end
+    ["NOT(#{conditions.join(' OR ')})", *assignments]
+  end
```
Comment (ulferts) [inline app/models/queries/projects/filters/name_and_identifier_filter.rb]:
```
def concatenate_with_values(condition, concatenation) 
  conditions = []
  assignments = []
  values.each do |value|
    conditions << condition
    assignments += [yield(value), yield(value)]
  end

  [conditions.join(" #{concatenation} "), *assignments]
end

# values is an instance variable and as such does not have to be provided
def where_equal
  concatenate_with_values('LOWER(projects.identifier) = ? OR LOWER(projects.name) = ?', 'OR') do |value|
    value.downcase
  end
end

...

def where_contains
  concatenate_with_values('LOWER(projects.identifier) LIKE ? OR LOWER(projects.name) LIKE ?', 'OR') do |value|
    "%#{value.downcase}%"
  end
end

def where_not_contains
  conditions = where_contains
  conditions[0] = "NOT(#{conditions[0]})"
  conditions
end

File: app/controllers/projects_controller.rb [hunk 14]
```diff
@@ -298,21 +297,30 @@ def add_current_user_to_project_if_not_admin(project)
...
+    if @query.valid?
+      # TODO: Use our query-based orders and pagination, instead of using this second version here.
+
+      projects = @query
+        .results
+        .with_required_storage
+        .with_latest_activity
+        .order(sort_clause)
+        .page(page_param)
+        .per_page(per_page_param)
```
Comment (ulferts) [inline app/controllers/projects_controller.rb]: adding
```
        .includes(:custom_values,
                  :enabled_modules)
```
will prevent n+1 queries I have seen in the console

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/c3c4854f8aed68a3a64c9be2b667482746036515) (deployed commit: b2f22deb02ec4a897bbe00b908ba3fff2ea7695f).

**Review by ulferts — COMMENTED** [review]
I haven't played around with the new code extensively, yet. 
Additionally, I haven't taken a look into the view as you wanted to alter it first.

I'v written down my specific code remarks inline.

What baffled me as a user was that the status filter can not be removed. It makes sense to have it as a default but preventing it from being removed is weird. I am aware that I can simply select all values to not filter at all but being prevented from removal is rude.

**Comment by wielinde** [comment]
Hey @lindenthal,

what do you think about @ulferts's concern:

> What baffled me as a user was that the status filter can not be removed. It makes sense to have it as a default but preventing it from being removed is weird. I am aware that I can simply select all values to not filter at all but being prevented from removal is rude.

**Comment by lindenthal** [comment]
@wielinde how should I answer this question? 

- Ifit takes one hour: do it now
- If it takes more than one hour: we do it later

Does this help?

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/689e33c6ec9e65bc5b1c2b2dabc4b9e439f74ac4) (deployed commit: 57d85756c30a74118b2839e99dbe5d89a11173e3).

**Comment by wielinde** [comment]
@lindenthal Well, I thought, disallowing to remove the active-or-archived filter was a requirement that came from you... Apparently I remembered it wrong.

**Comment by lindenthal** [comment]
@wielinde 

No, not that I am aware of.

Best
Niels

File: app/models/queries/filters/strategies/date_time_past.rb (hunk 11 above)
Comment (wielinde) [reply app/models/queries/filters/strategies/date_time_past.rb]: ... I ran into troubles setting the `default_operator` here when it was already set in for the integer strategy. So I tried to make it "work" somehow...

File: app/helpers/sort_helper.rb (hunk 9 above)
Comment (wielinde) [reply app/helpers/sort_helper.rb]: Very good point! I hope to have that resolved with https://github.com/opf/openproject/pull/5981/commits/65e69415eefd7eafa48b5249a587d7f48438ae0b

File: app/helpers/projects_helper.rb (hunk 8 above)
Comment (wielinde) [reply app/helpers/projects_helper.rb]: @ulferts I agree with you, that the visibility of filters should be determined within the query and not within the projects helper. Unfortunately I am stuck here for two reasons:

- I don't know the visibility requirement for `AncestorFilter`. It is not part of the project table. Where does it get used? And what are the requirements there for visibility?
- Where and how would I put the code you wrote above? I tried `Queries::Projects` but wasn't successful.

Could you lend me a hand?

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/2895db3e43500c5fcb8bda29367f1bec0c1b9006) (deployed commit: 80c559c7e8bda7b8fc413bd4c0bf2ba06c886780).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/e8da2d45dc04ba4086f22dfe1b7d82d497bc3bbe) (deployed commit: d8ec874dc1359c4f364c6e1fb2ec57473094d0cb).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b379a6df24969c7a69619c82272189023dba5cc1) (deployed commit: 153335746e48103a6887d7d311a6336becdabbb7).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/7262273fa3847017e9e0f70b16491ac687a74aff) (deployed commit: d7dc2a6c836fce477b8cd2af458769f17054dc57).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/0ba8387a5ee9db507267f7d04561df5852ed2106) (deployed commit: eb6d2662238d9987074562b1aabe9d93e05e4f39).

**Comment by lindenthal** [comment]
Hi @ulferts  und @wielinde 

image

- The indent of the project description looks akward. Is there a specific reason for this?
- Can we use the existing breadcrumb icon to show the subproject?

Best
Niels

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/ff98131b3902b8754debe0ee29c36e3c40340900) (deployed commit: 75695aca3f3f4c257905447f86ae0f8602b58b4d).

**Comment by ulferts** [comment]
@lindenthal 

> The indent of the project description looks akward. Is there a specific reason for this?

Fixed in 75695ac.

> Can we use the existing breadcrumb icon to show the subproject?

Where do you want to have the icon used. On the parent as is done in the wp table or on the child level?

Replacing the icon on the child level is trivial. However, it would be inconsistent compared to the wp table. Adapting the behaviour to be equal to the wp table would require rewriting the way the hierarchy is calculated which is currently done by code shared with a couple of places (more than one less than four hours).

As it is, I wouldn't recommend doing it.

**Comment by lindenthal** [comment]
Hi @ulferts,

yes, let‘s simply change the back on the child level.

Thanks
Niels

**Comment by ulferts** [comment]
Done in a8b2067 @lindenthal and I don't like the results:

image

To users used to the wp table, the icon suggests being able to click on it to expand/collapse the children and ancestors of the project as this is how it is done on the wp table:
image
and actually on the projects table as well (expand/collapse of the project description)
image

Can we maybe use a different icon until we get around to implement the ability to expand/collapse project hierarchies?

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/81dbdd1be20c06d22a38d14e2b5ca60e5723f616) (deployed commit: a8b2067f23ce27b217153a9eee448072cf9f5809).

**Comment by lindenthal** [comment]
Hi @ulferts

I agree. That is why I proposed to take the icons from the breadcrumb.

Best
Niels

**Comment by ulferts** [comment]
@lindenthal, sorry I misread. Way better now:

image

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/143f9a13beed02d970d6ba8d1ff789772a276640) (deployed commit: 891e8d46343ea202b18f6fc3361dfb7bdc17adfd).

**Comment by ulferts** [comment]
Please note that the spec failing is the invalid_query feature spec which is prone to fail. The test works just fine locally.

File: app/helpers/projects_helper.rb [hunk 15]
```diff
@@ -129,42 +125,147 @@ def filter_set?
...
+
+  def deactivate_class_on_lft_sort
+    if sorted_by_lft?
+      '-inactive'
+    end
+  end
+
+  def href_only_when_not_sort_lft
+    unless sorted_by_lft?
+      "href=#{projects_path(sortBy: JSON::dump([['lft', 'asc']]))}"
```
Comment (wielinde) [inline app/helpers/projects_helper.rb]: This resets filters unexpectedly.

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/b114d1ef4bbb2f62e41ef01db08da79d73a7728c) (deployed commit: 1f01b1ad5f50a91b046c9aae8b92512cc3359c63).

File: app/models/queries/filters/strategies/cf_float.rb [hunk 16]
```diff
@@ -28,17 +29,9 @@
...
-      super_value = super.dup
-      super_value['!*'] = Queries::Operators::NoneOrBlank
-      super_value['*'] = Queries::Operators::AllAndNonBlank
-      super_value['>='] = Queries::Operators::CastedGreaterOrEqual
-      super_value['<='] = Queries::Operators::CastedLessOrEqual
-
-      super_value
-    end
+  class CfFloat < BaseStrategy
+    include Queries::Filters::Strategies::Numeric
```
Comment (wielinde) [inline app/models/queries/filters/strategies/cf_float.rb]: The idea is that one include can override the other?

File: app/models/queries/projects.rb [hunk 17]
```diff
@@ -38,4 +40,9 @@ module Queries::Projects
   register.filter query, filters::CustomFieldFilter
   register.filter query, filters::CreatedOnFilter
   register.filter query, filters::LatestActivityAtFilter
+
+  register.order query, orders::DefaultOrder
+  register.order query, orders::LatestActivityAtOrder
+  register.order query, orders::RequiredDiskSpaceOrder
+  register.order query, orders::CustomFieldOrder
```
Comment (wielinde) [inline app/models/queries/projects.rb]: Why does ordering of "name" or "created_on" work?

File: app/models/queries/projects/filters/custom_field_filter.rb [hunk 18]
```diff
@@ -30,184 +30,43 @@
...
+  self.custom_field_class = ProjectCustomField
 
   def type
-    return nil unless custom_field
-
-    case custom_field.field_format
-    when 'int'
-      :integer
-    when 'float'
+    if custom_field && custom_field.field_format == 'float'
```
Comment (wielinde) [inline app/models/queries/projects/filters/custom_field_filter.rb]: This extra treatment for project custom fields of type float might not be necessary, as I believe, that the for work packages CFs of type float do not work as expected neither. So maybe we could solve that there too...

File: app/models/queries/projects/orders/custom_field_order.rb [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+  def custom_field
+    @custom_field ||= begin
+      id = self.class.key.match(attribute)[1]
+
+      ProjectCustomField.visible.find_by_id(id)
+    end
+  end
+
+  def scope
+    super.select(custom_field.order_statements)
```
Comment (wielinde) [inline app/models/queries/projects/orders/custom_field_order.rb]: Why is this a `select`?

File: spec/features/projects/projects_index_spec.rb [hunk 20]
```diff
@@ -124,14 +140,14 @@ def remove_filter(name)
...
+        Role.anonymous
         visit projects_path
 
         expect(page).to_not have_text(project.name)
         expect(page).to have_text(public_project.name)
 
         # Test that the 'More' menu stays invisible on hover
-        page.find('tbody tr').hover
-        expect(page).to_not have_css('.icon-show-more-horizontal')
+        expect(page).to_not have_selector('.icon-show-more-horizontal')
```
Comment (wielinde) [inline spec/features/projects/projects_index_spec.rb]: Hmm, is this still the same? Don't we need to ensure that we hovered first so that we can be assured that the menu does not show up then??

UPDATE: Just saw your comment further down about opacity...

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/042b6bb0d35549af8698e82fc8f88088a6f043c6) (deployed commit: 11fddf9985c1874aa97f6ddf4e0216e5add1291e).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/c2a2599dcbbf565de06677e3fdc7363f913a6d11) (deployed commit: 43e17f18eb1aa31c2d118ea3a1e77097f2561939).

**Comment by crohr** [comment]
A new preview environment is ready! [Click to open](https://pullpreview.com/opf/openproject/sessions/453f5907fba67bf465ec3b842a90a50841870d28) (deployed commit: 85a6522628d200618f344fa505bd0dcdedcfc1f7).

**Comment by ulferts** [comment]
👍
